Gas Collection

CPhuong24 opened the pull request:
Added a few lines for gas collection and reserving gas collectors. 

Changed build order slightly to include Assimilator (Refinery).

CPhuong24 committed:
Gas Collection

Added a few lines to reserve gas collectors and the collection of the gas resource

CPhuong24 committed:
Changed StrategyManager Build Order

Changed build order to include Assimilator to collect gas

Note: For some reason putting the assimilator after the Gateway in the build order causes the bot to stall and skip straight to attacking without building the Gateway

CPhuong24 committed:
Update StarterBot_Config_Unused.txt

removed auto_menu line

CPhuong24 committed:
Gas Collection Comments

Added a few lines to comment the Gas Collection Code

CPhuong24 committed:
Update to run "StarterBot.exe"

VincentTruong91 commented on src/starterbot/StrategyManager.cpp:
```diff
@@ -24,10 +24,11 @@ void StrategyManager::onStart() {
         {ActionType::TRAIN,  ActionItem::NONE, bw::UnitTypes::Protoss_Probe,        7 }, // 0
         {ActionType::BUILD,  0,                bw::UnitTypes::Protoss_Pylon,        1 }, // 1
         {ActionType::SCOUT,  1,                bw::UnitTypes::Protoss_Probe,        1 }, // 2
+        {ActionType::BUILD,  1,                bw::UnitTypes::Protoss_Assimilator,  1 }, // 4
```
Comments should be `3`, then `4` instead of `4`, `3`, nothing too crazy though.

VincentTruong91 reviewed (approved):
Everything else besides the comment looks good to me, I approve.

v-rob commented on src/starterbot/ProductionManager.h:
```diff
@@ -13,6 +13,8 @@ class ProductionManager : public EventReceiver {
     // gathering minerals are not reserved to keep them available for other managers.
     bw::Unitset m_workers;
 
+    bw::Unitset m_gas_gatherers;
```
Should be `m_gasGatherers`, since we use `camelCase` for variable names. The underscore is only used for the `m_` prefix.

v-rob commented on src/starterbot/ProductionManager.cpp:
```diff
@@ -60,6 +60,29 @@ void ProductionManager::onFrame() {
+    // Check to see if refinery exists
+    for (bw::Unit unit : m_unitManager.borrowUnits(bw::Filter::IsRefinery)) {
+        if (m_gas_gatherers.size() < 2) {
+            m_gas_gatherers = m_unitManager.reserveUnits(bw::Filter::IsWorker, 2);
```
Subtle bug here. If `m_gas_gatherers` has one unit in it, then you reserve two new units, and the old unit is thrown away without being freed. So you've effectively got a unit that has disappeared from `UnitManager` and can never be used again.

You should probably do
```cpp
if (m_gas_gatherers.size() < 2) {
    bw::Unitset newGatherers = m_unitManager.reserveUnits(bw::Filter::IsWorker, 2 - m_gas_gatherers.size());
    m_gas_gatherers.insert(newGatherers.begin(), newGatherers.end());
}
```

v-rob commented on src/starterbot/ProductionManager.cpp:
```diff
@@ -60,6 +60,29 @@ void ProductionManager::onFrame() {
     m_unitManager.releaseUnits(m_workers, bw::Filter::IsIdle);
 
+    // Check to see if refinery exists
+    for (bw::Unit unit : m_unitManager.borrowUnits(bw::Filter::IsRefinery)) {
```
Why do you loop over every refinery? `m_gas_gatherers` will never have more than two units at any time. Did you mean to have two units per refinery, or just two units total?

If you just need two units total, then you can do
```cpp
if (m_unitManager.borrowUnit(bw::Filter::IsRefinery) != nullptr)
```
rather than a for loop. If you need two units per refinery, then you should probably do
```cpp
int gasCount = m_unitManager.borrowCount(bw::Filter::IsRefinery) * 2;
if (m_gas_gatherers.size() < gasCount) {
    bw::Unitset newGatherers = m_unitManager.reserveUnits(bw::Filter::IsWorker, gasCount - m_gas_gatherers.size());
    m_gas_gatherers.insert(newGatherers.begin(), newGatherers.end());
}
```

v-rob commented on src/starterbot/StrategyManager.cpp:
```diff
@@ -24,10 +24,11 @@ void StrategyManager::onStart() {
-        {ActionType::ATTACK, 4,                                                       }, // 6
+        {ActionType::TRAIN,  3,                bw::UnitTypes::Protoss_Zealot,       2 }, // 5
+        {ActionType::TRAIN,  3,                bw::UnitTypes::Protoss_Probe,        13}, // 6
+        {ActionType::ATTACK, 4,                                                       }, // 7
```
Should be `4, 4, 5` rather than `3, 3, 4` since you've added a new dependency into the chain.

CPhuong24 committed:
Finished Gas collection unit assignment

Followed comments from branch and tweaked code to better allocate gas collection workers.

Commit:
Make gas collection balanced and add dragoon creation
